### ISSUE TITLE
Add KFAC optimizer

### ISSUE THREAD

**Issue by n-gao** [pytorch/pytorch#35801]
## 🚀 Feature
Add the [Kronecker-factored Approximate Curvature (KFAC)](https://arxiv.org/abs/1503.05671) optimizer and/or [E-KFAC](https://arxiv.org/abs/1806.03884).

## Motivation
These are both approximate second order methods with promising results in terms of convergence rate. KFAC uses an block-diagonal approximation to the Fisher information matrix to approximate natural gradients. This allows for larger learning rates and thus improved convergence rate. Having a fully features built-in optimizer allows researchers and deep learning engineers to quickly iterate and test.

## Pitch
A KFAC implementation that works out of the box for all supported layers should be added to the built-in optimizers.

## Additional context
[An unmaintained Pytorch implementation](https://github.com/alecwangcq/KFAC-Pytorch)
[The standard TF implementation](https://github.com/tensorflow/kfac)
[KFAC for CNN](https://arxiv.org/abs/1602.01407)
[KFAC for RNN](https://openreview.net/forum?id=HyMTkQZAb)

**Comment by gchanan** [comment]
@vincentqb can you make a call on if we want this in PyTorch core?

**Comment by n-gao** [comment]
I want to add two figures to illustrate why KFAC is, imo, such an important algorithm. The figures show the optimization landscape and how KFAC and SGD reach the optimum.
![image](https://user-images.githubusercontent.com/9084670/81046647-22abcd80-8eb9-11ea-8512-0adb424083eb.png)
![image](https://user-images.githubusercontent.com/9084670/81046698-3e16d880-8eb9-11ea-9291-423117e1ab11.png)

**Comment by n-gao** [comment]
In case someone is looking for a KFAC implementation in KFAC: https://github.com/n-gao/pytorch-kfac
